monkey/monkey_island/cc/repository/mongo_machine_repository.py

```diff
@@ -0,0 +1,102 @@
+from ipaddress import IPv4Address
+from threading import Lock
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from common.types import HardwareID
+from monkey_island.cc.models import Machine, MachineID
+
+from . import IMachineRepository, RemovalError, RetrievalError, StorageError, UnknownRecordError
+from .consts import MONGO_OBJECT_ID_KEY
+
+
+class MongoMachineRepository(IMachineRepository):
+    """A repository used to store and retrieve Machines in MongoDB"""
+
+    def __init__(self, mongo_client: MongoClient):
+        self._machines_collection = mongo_client.monkey_island.machines
+        self._id_lock = Lock()
+        self._next_id = self._get_biggest_id()
+
+    def _get_biggest_id(self) -> MachineID:
+        try:
+            return self._machines_collection.find().sort("id", -1).limit(1)[0]["id"]
+        except IndexError:
+            return 0
```

Comment:
Mongo `ObjectIDs` are generated on the client side, which is actually convenient. Mongo `ObjectIDs` can also be easily converted to `int`, which is also convenient.

ObjectIDs are 12 bytes, which is larger than necessary for our purposes. Mongo creates them anyway, but our model doesn't need to hold onto them. ObjectIDs may also take slightly more processing to manipulate. This very likely won't have any real world effect on performance either way.

Using `ObjectIDs` would involve some extra logic in serializing/deserializing. Probably roughly the same amount of logic to manage sequential IDs.

The way I see it, switching to using the `ObjectIDs` doesn't really help us gain anything over sequential IDs. Since sequential IDs are already implemented, I say we just leave them for now.